server/server_test.go

```diff
@@ -228,6 +231,156 @@ func TestErrorInstrumentationMiddleware(t *testing.T) {
+		_, _ = w.Write([]byte("OK"))
+	})
+	server.HTTP.HandleFunc("/error500", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(500)
+	})
+	server.HTTP.HandleFunc("/sleep10", func(w http.ResponseWriter, r *http.Request) {
+		_, _ = io.Copy(ioutil.Discard, r.Body) // Consume body, otherwise it's not counted.
+		_ = cancelableSleep(r.Context(), time.Second*10)
+	})
+
+	go server.Run()
+
+	callThenCancel := func(f func(ctx context.Context) error) error {
+		ctx, cancel := context.WithCancel(context.Background())
+		errChan := make(chan error, 1)
+		go func() {
+			errChan <- f(ctx)
+		}()
+		time.Sleep(50 * time.Millisecond) // allow the call to reach the handler
+		cancel()
+		return <-errChan
+	}
+
+	// Now test the HTTP versions of the functions
+	{
+		req, err := http.NewRequest("GET", "http://127.0.0.1:9090/succeed", nil)
+		require.NoError(t, err)
+		resp, err := http.DefaultClient.Do(req)
+		require.NoError(t, err)
+		body, err := ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		require.Equal(t, "OK", string(body))
+	}
+	{
+		req, err := http.NewRequest("GET", "http://127.0.0.1:9090/error500", nil)
+		require.NoError(t, err)
+		_, err = http.DefaultClient.Do(req)
+		require.NoError(t, err)
+	}
+	{
+		req, err := http.NewRequest("POST", "http://127.0.0.1:9090/sleep10", bytes.NewReader([]byte("Body")))
+		require.NoError(t, err)
+		err = callThenCancel(func(ctx context.Context) error {
+			_, err = http.DefaultClient.Do(req.WithContext(ctx))
+			return err
+		})
+		require.Error(t, err, context.Canceled)
+	}
+
+	server.Shutdown()
+
+	require.NoError(t, testutil.GatherAndCompare(prometheus.DefaultGatherer, bytes.NewBufferString(`
+		# HELP inflight_requests Current number of inflight requests.
+		# TYPE inflight_requests gauge
+		inflight_requests{method="POST",route="sleep10"} 0
+		inflight_requests{method="GET",route="succeed"} 0
+       	inflight_requests{method="GET",route="error500"} 0
+
+		# HELP request_message_bytes Size (in bytes) of messages received in the request.
+		# TYPE request_message_bytes histogram
+		request_message_bytes_bucket{method="GET",route="error500",le="1.048576e+06"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="2.62144e+06"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="5.24288e+06"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="1.048576e+07"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="2.62144e+07"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="5.24288e+07"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="1.048576e+08"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="2.62144e+08"} 1
+		request_message_bytes_bucket{method="GET",route="error500",le="+Inf"} 1
+		request_message_bytes_sum{method="GET",route="error500"} 0
+		request_message_bytes_count{method="GET",route="error500"} 1
```

Comment:
Not your fault, but this is ugly as heck.  Suggest `prometheus/testutil` could be augmented with a way to provide expected histogram counts.

Reply:
We can make it slightly more beautiful by not using scientific notation in bucket sizes. Not sure if that would help much though.

Reply:
Is that something you can control?  

I was thinking something like `testutil.ExpectedHistogram(inflightRequests, "GET", "error500", 100)`

Reply:
We can control format of the floats in our text output, yet.

As somewhat nicer but still ugly solution, I'm thinking
```go
		# HELP request_message_bytes 
		# TYPE request_message_bytes histogram`+
		histogram([]float64{0}, middleware.BodySizeBuckets, "request_message_bytes_bucket", "method", "GET", "route", "error500")+
		histogram([]float64{4}, middleware.BodySizeBuckets, "request_message_bytes_bucket", "method", "GET", "route", "sleep10")+
		histogram([]float64{0}, middleware.BodySizeBuckets, "request_message_bytes_bucket", "method", "GET", "route", "succeed")+...
```

with `histogram` function to generate the text.

Your proposed `testutil.ExpectedHistogram` is an option too. We can try it in this lib first, and if it is nice, can try to send to Prometheus client library.